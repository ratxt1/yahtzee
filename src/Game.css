.Game-reroll {
  margin-top: 1em;
  font-size: 1.5em;
  border-radius: 0.2em;
}

.Game-reroll:hover {
  background-color: pink;
}

.Game-play-again {
  font-size: 1.5em;
  border-radius: 0.2em;
}

.Game-play-again:hover {
  background-color: greenyellow;
}