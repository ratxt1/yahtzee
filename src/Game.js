import React, { Component } from 'react';
import Dice from './Dice';
import ScoreTable from './ScoreTable';
import './Game.css';

const NUM_DICE = 5;
const NUM_ROLLS = 3;

class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {
      dice: Array
        .from({ length: NUM_DICE })
        .map(d => Math.ceil(Math.random() * 6)),
      locked: Array(NUM_DICE).fill(false),
      rollsLeft: NUM_ROLLS - 1,
      scores: {
        ones: undefined,
        twos: undefined,
        threes: undefined,
        fours: undefined,
        fives: undefined,
        sixes: undefined,
        threeOfKind: undefined,
        fourOfKind: undefined,
        fullHouse: undefined,
        smallStraight: undefined,
        largeStraight: undefined,
        yahtzee: undefined,
        chance: undefined
      },
      diceRolling: false
    };
    this.roll = this.roll.bind(this);
    this.doScore = this.doScore.bind(this);
    this.toggleLocked = this.toggleLocked.bind(this);
    this.handleDieRoll = this.handleDieRoll.bind(this);
    this.reset = this.reset.bind(this);
  }

  roll(evt) {
    // roll dice whose indexes are in reroll
    this.setState(st => ({
      dice: st.dice.map(
        (d, i) => st.locked[i] ? d : Math.ceil(Math.random() * 6)),
      locked: st.rollsLeft > 1 ? st.locked : Array(NUM_DICE).fill(true),
      rollsLeft: st.rollsLeft - 1,
      diceRolling: false
    }));
  }

  handleDieRoll(evt) {
    // toggle class to shake die on roll
    this.setState(st => ({
      diceRolling: true
    }));
    setTimeout(this.roll, 500)
  }

  toggleLocked(idx) {
    // toggle whether idx is in locked or not
    this.setState(st => ({
      locked: [
        ...st.locked.slice(0, idx),
        !st.locked[idx],
        ...st.locked.slice(idx + 1)
      ],
    }))
  }

  reset() {
    this.setState({
      dice: Array
        .from({ length: NUM_DICE })
        .map(d => Math.ceil(Math.random() * 6)),
      locked: Array(NUM_DICE).fill(false),
      rollsLeft: NUM_ROLLS - 1,
      scores: {
        ones: undefined,
        twos: undefined,
        threes: undefined,
        fours: undefined,
        fives: undefined,
        sixes: undefined,
        threeOfKind: undefined,
        fourOfKind: undefined,
        fullHouse: undefined,
        smallStraight: undefined,
        largeStraight: undefined,
        yahtzee: undefined,
        chance: undefined
      },
      diceRolling: false
    })
  }

  doScore(rulename, ruleFn) {
    // evaluate this ruleFn with the dice and score this rulename
    this.setState(st => ({
      scores: { ...st.scores, [rulename]: ruleFn(this.state.dice) },
      rollsLeft: NUM_ROLLS,
      locked: Array(NUM_DICE).fill(false),
    }));
    this.handleDieRoll();
  }

  render() {
    return (
      <section>
        <div className={this.state.diceRolling ? "Die-shake": ""}>
          <Dice dice={this.state.dice}
            locked={this.state.locked}
            toggleLocked={!this.state.locked.every(x => x) ? this.toggleLocked : () => { }}
          />
        </div>

        <button
          className="Game-reroll"
          disabled={this.state.locked.every(x => x)}
          onClick={this.handleDieRoll}>
          {this.state.rollsLeft} Rerolls Left
        </button>
        <ScoreTable doScore={this.doScore} scores={this.state.scores} diceRolling={this.state.diceRolling} />
        {(Object.values(this.state.scores).every(s => s !== undefined)) ? <button onClick={this.reset}>Play Again?</button> : null}
      </section >
    );
  }
}

export default Game;