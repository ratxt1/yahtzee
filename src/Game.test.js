import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import Game from "./Game";


it("renders without crashing", function() {
  shallow(<Game />);
});

it("matches snapshot", function() {
  let wrapper = shallow(<Game />);
  let serialized = toJson(wrapper);
  expect(serialized).toMatchSnapshot();
});

it("can lock a die on click", function() {
  let wrapper = mount(<Game />);
  const die = wrapper.find(".Die").first();
  die.simulate("click");

  expect(wrapper.state().locked[0]).toEqual(true);
});

it("can't roll if rollsLeft is equal to 0", function() {
  let wrapper = mount(<Game />);
  wrapper.setState({ rollsLeft: 0 });
  wrapper.instance().roll();
 
  expect(wrapper.state().rollsLeft).toEqual(0);
});

it("can't reassign score in scoretable", function() {
  let wrapper = mount(<Game />);
  let score = wrapper.find('.RuleRow').first();
  expect(score.hasClass("RuleRow-active")).toBe(true);

  score.simulate('click');

  let currentScore = wrapper.state().scores.ones
  expect(currentScore).not.toBe(undefined);

  score.simulate('click');
  expect(wrapper.state().scores.ones).toBe(currentScore)

  score = wrapper.find('.RuleRow').first();
  expect(score.hasClass("RuleRow-active")).toBe(false);

});