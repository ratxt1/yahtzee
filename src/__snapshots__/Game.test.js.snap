// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<section>
  <Dice
    dice={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
    locked={
      Array [
        false,
        false,
        false,
        false,
        false,
      ]
    }
    toggleLocked={[Function]}
  />
  <button
    className="Game-reroll"
    disabled={false}
    onClick={[Function]}
  >
    3
     Rerolls Left
  </button>
  <ScoreTable
    doScore={[Function]}
    scores={
      Object {
        "chance": undefined,
        "fives": undefined,
        "fourOfKind": undefined,
        "fours": undefined,
        "fullHouse": undefined,
        "largeStraight": undefined,
        "ones": undefined,
        "sixes": undefined,
        "smallStraight": undefined,
        "threeOfKind": undefined,
        "threes": undefined,
        "twos": undefined,
        "yahtzee": undefined,
      }
    }
  />
</section>
`;
